master/buildbot/reporters/telegram.py

```diff
@@ -0,0 +1,1055 @@
+        if user is None:
+            self.log('no user in incoming telegram message')
+            return defer.succeed('no user')
+
+        chat = message.get('chat')
+        if chat is None:
+            self.log('no chat in telegram message')
+            return defer.succeed('no chat in the message')
+
+        text = message.get('text')
+        if not text:
+            return defer.succeed('no text in the message')
+
+        contact = self.getContact(user=user, channel=chat)
+        data['tmessage'] = message
+        template, contact.template = contact.template, None
+        if text.startswith(self.commandPrefix):
+            d = contact.handleMessage(text, **data)
+        else:
+            if template:
+                text = template.format(shlex.quote(text))
+            d = contact.handleMessage(text, **data)
+        return d
+
+    @defer.inlineCallbacks
+    def _post(self, path, **kwargs):
+        try:
+            res = yield self.http_client.post(path, **kwargs)
+            ans = yield res.json()
+            if not ans.get('ok'):
+                raise ValueError("[{}] {}".format(res.code, ans.get('description')))
+        except AssertionError as err:
+            # just for tests
+            raise err
+        except Exception as err:
+            self.log("ERROR: cannot send telegram request {}: {}".format(path, err))
+        else:
+            return ans.get('result', True)
+
+    @defer.inlineCallbacks
+    def _post_until_success(self, path, **kwargs):
+        logme = True
+        while True:
+            try:
+                res = yield self.http_client.post(path, **kwargs)
+                ans = yield res.json()
+                if not ans.get('ok'):
+                    self.log("ERROR: cannot send telegram request {}: "
+                             "[{}] {}".format(path, res.code, ans.get('description')))
+                    return
+            except AssertionError as err:
+                # just for tests
+                raise err
+            except Exception as err:
+                msg = "ERROR: cannot send telegram request {} (will try again): {}".format(path, err)
+                if logme:
+                    self.log(msg)
+                    logme = False
+                yield task.deferLater(self.reactor, self.retry_delay, lambda: None)
+            else:
+                return ans.get('result', True)
+
+    @defer.inlineCallbacks
+    def set_nickname(self):
+        res = yield self._post_until_success('/getMe')
+        if res:
+            self.nickname = res.get('username')
+
+    @defer.inlineCallbacks
+    def answer_query(self, query_id, notify=None):
+        params = dict(callback_query_id=query_id)
+        if notify is not None:
+            params.update(dict(text=notify))
+        return (yield self._post('/answerCallbackQuery', json=params))
+
+    @defer.inlineCallbacks
+    def send_message(self, channel, message, parse_mode='Markdown', **kwargs):
```

Comment:
Could we require `channel` to be `TelegramChannel`?

Reply:
I have changed the name of the argument to `chat`, which is numeric chat id. This is consistent with other methods of this kind.